File: mcp_server/src/codebase_indexer/indexer/graph_writer.py
"""Index-time Neo4j graph writer (ADR 0002 Phase 1)."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

import structlog

from codebase_indexer.indexer.chunker import Chunk, extract_imported_names
from codebase_indexer.tools.build_deps import (
    extract_build_deps,
    is_build_manifest,
    match_deps_to_collections,
)
from codebase_indexer.tools.cross_references import UrlExtractors

log = structlog.get_logger()

_HTTP_METHOD_PATTERN = re.compile(
    r"@(?:Get|Post|Put|Delete|Patch|Request)Mapping|"
    r"\[(?:Http(?:Get|Post|Put|Delete|Patch))\]|"
    r"\.Map(Get|Post|Put|Delete|Patch)\(|"
    r"(?:app|router)\.(get|post|put|delete|patch)\(",
    re.IGNORECASE,
)
_CONFIG_FILE_PATTERN = re.compile(
    r"\.(ya?ml|json|properties|env|config)$", re.IGNORECASE
)


def symbol_qualified_name(collection: str, rel_path: str, symbol_name: str) -> str:
    """Stable Symbol key: {collection}:{rel_path}::{symbol_name}."""
    return f"{collection}:{rel_path}::{symbol_name}"


def import_qualified_name(collection: str, import_name: str) -> str:
    """Qualified name for an unresolved import target."""
    return f"{collection}::import::{import_name}"


def callee_qualified_name(collection: str, callee: str) -> str:
    """Qualified name for a best-effort callee symbol."""
    return f"{collection}::callee::{callee}"


def artifact_key(group: str, name: str, ecosystem: str) -> str:
    """Stable Artifact key across collections."""
    return f"{ecosystem}:{group}:{name}" if group else f"{ecosystem}:{name}"


def infer_http_method(content: str) -> str:
    """Best-effort HTTP method inference from endpoint definition content."""
    match = _HTTP_METHOD_PATTERN.search(content)
    if not match:
        return ""
    token = match.group(0).lower()
    for method in ("get", "post", "put", "delete", "patch"):
        if method in token:
            return method.upper()
    return ""


def extract_file_import_names(content: str, language: str) -> list[str]:
    """Collect imported symbol names from raw file content."""
    seen: set[str] = set()
    names: list[str] = []
    for line in content.splitlines():
        imported = extract_imported_names(line, language)
        if imported is None or not imported:
            continue
        for name in imported:
            if name not in seen:
                seen.add(name)
                names.append(name)
    return names


@dataclass
class GraphBatch:
    """Structured graph upsert payload for one flush batch."""

    collection: str
    schema_version: int = 1
    collection_props: bool = True
    files: list[dict] = field(default_factory=list)
    chunks: list[dict] = field(default_factory=list)
    defines: list[dict] = field(default_factory=list)
    calls: list[dict] = field(default_factory=list)
    imports: list[dict] = field(default_factory=list)
    endpoints: list[dict] = field(default_factory=list)
    declares_endpoint: list[dict] = field(default_factory=list)
    http_calls: list[dict] = field(default_factory=list)
    configures: list[dict] = field(default_factory=list)
    build_deps: list[dict] = field(default_factory=list)
    resolves_to: list[dict] = field(default_factory=list)


def build_graph_batch(
    *,
    collection: str,
    chunks: list[Chunk],
    url_extractors: UrlExtractors,
    workspace_path: str,
    collection_names: list[str],
    schema_version: int = 1,
) -> GraphBatch:
    """Build a graph batch from indexed chunks (grouped by file)."""
    batch = GraphBatch(collection=collection, schema_version=schema_version)
    if not chunks:
        return batch

    by_file: dict[str, list[Chunk]] = {}
    for chunk in chunks:
        by_file.setdefault(chunk.rel_path, []).append(chunk)

    seen_files: set[str] = set()
    seen_build_keys: set[str] = set()

    for rel_path, file_chunks in by_file.items():
        first = file_chunks[0]
        if rel_path not in seen_files:
            seen_files.add(rel_path)
            batch.files.append(
                {
                    "rel_path": rel_path,
                    "language": first.language,
                    "sha256": first.file_sha256,
                }
            )

        for chunk in file_chunks:
            batch.chunks.append(
                {
                    "chunk_id": chunk.chunk_id,
                    "rel_path": rel_path,
                    "start_line": chunk.start_line,
                    "end_line": chunk.end_line,
                }
            )

            if chunk.symbol_name:
                qn = symbol_qualified_name(collection, rel_path, chunk.symbol_name)
                batch.defines.append(
                    {
                        "chunk_id": chunk.chunk_id,
                        "qualified_name": qn,
                        "name": chunk.symbol_name,
                        "kind": chunk.symbol_type,
                    }
                )

            for callee in chunk.callees:
                batch.calls.append(
                    {
                        "chunk_id": chunk.chunk_id,
                        "qualified_name": callee_qualified_name(collection, callee),
                        "name": callee,
                    }
                )

            for route in url_extractors.route_paths(chunk.content, rel_path):
                method = infer_http_method(chunk.content)
                batch.endpoints.append({"path": route, "method": method})
                batch.declares_endpoint.append(
                    {"chunk_id": chunk.chunk_id, "path": route}
                )

            for path in url_extractors.code_urls(chunk.content):
                batch.endpoints.append({"path": path, "method": ""})
                batch.http_calls.append({"chunk_id": chunk.chunk_id, "path": path})

            if _CONFIG_FILE_PATTERN.search(rel_path):
                config_paths, _base_urls = url_extractors.config_urls(chunk.content)
                for path in config_paths:
                    batch.endpoints.append({"path": path, "method": ""})
                    batch.configures.append({"chunk_id": chunk.chunk_id, "path": path})

        file_content = _read_file_content(workspace_path, rel_path)
        if file_content is None:
            file_content = "\n".join(c.content for c in file_chunks)

        for import_name in extract_file_import_names(file_content, first.language):
            batch.imports.append(
                {
                    "rel_path": rel_path,
                    "qualified_name": import_qualified_name(collection, import_name),
                    "name": import_name,
                }
            )

        if is_build_manifest(rel_path):
            manifest_content = _read_file_content(workspace_path, rel_path) or file_content
            deps = extract_build_deps(manifest_content, rel_path)
            matches = match_deps_to_collections(
                deps,
                collection_names,
                self_collection=collection,
            )
            match_by_artifact: dict[str, str] = {
                m["artifact"]: m["matched_collection"] for m in matches
            }
            for dep in deps:
                key = artifact_key(dep.group, dep.artifact, dep.ecosystem)
                if key in seen_build_keys:
                    continue
                seen_build_keys.add(key)
                batch.build_deps.append(
                    {
                        "key": key,
                        "name": dep.artifact,
                        "group": dep.group,
                        "ecosystem": dep.ecosystem,
                        "version": dep.version,
                        "scope": dep.scope,
                    }
                )
                target = match_by_artifact.get(dep.artifact)
                if target:
                    batch.resolves_to.append(
                        {
                            "artifact_key": key,
                            "target_collection": target,
                        }
                    )

    return batch


def _read_file_content(workspace_path: str, rel_path: str) -> str | None:
    """Re-read manifest/source from disk for full-file extractors."""
    path = Path(workspace_path) / rel_path.replace("\\", "/")
    try:
        return path.read_text(encoding="utf-8", errors="replace")
    except OSError as exc:
        log.warning("graph_writer_read_failed", path=rel_path, error=str(exc))
        return None


async def write_chunks_to_graph(
    graph_storage,
    *,
    collection: str,
    chunks: list[Chunk],
    url_extractors: UrlExtractors,
    workspace_path: str,
    collection_names: list[str],
    schema_version: int = 1,
) -> None:
    """Build and persist a graph batch (no-op when storage disabled)."""
    if graph_storage is None or not graph_storage.enabled:
        return

    batch = build_graph_batch(
        collection=collection,
        chunks=chunks,
        url_extractors=url_extractors,
        workspace_path=workspace_path,
        collection_names=collection_names,
        schema_version=schema_version,
    )
    await graph_storage.write_batch(batch)

File: src/CodebaseIndexer.Application/Graph/GraphWriter.cs
using System.Text.RegularExpressions;
using CodebaseIndexer.Application.BuildDeps;
using CodebaseIndexer.Application.Search;
using CodebaseIndexer.Domain.Models;
using Microsoft.Extensions.Logging;

namespace CodebaseIndexer.Application.Graph;

/// <summary>Index-time Neo4j graph batch builder (Python <c>graph_writer.py</c>).</summary>
public sealed partial class GraphWriter
{
    private readonly ILogger<GraphWriter> _logger;

    /// <summary>Creates the graph writer.</summary>
    public GraphWriter(ILogger<GraphWriter> logger) => _logger = logger;

    /// <summary>Stable Symbol key: {collection}:{rel_path}::{symbol_name}.</summary>
    public static string SymbolQualifiedName(string collection, string relPath, string symbolName) =>
        $"{collection}:{relPath}::{symbolName}";

    /// <summary>Qualified name for an unresolved import target.</summary>
    public static string ImportQualifiedName(string collection, string importName) =>
        $"{collection}::import::{importName}";

    /// <summary>Qualified name for a best-effort callee symbol.</summary>
    public static string CalleeQualifiedName(string collection, string callee) =>
        $"{collection}::callee::{callee}";

    /// <summary>Stable Artifact key across collections.</summary>
    public static string ArtifactKey(string group, string name, string ecosystem) =>
        string.IsNullOrEmpty(group) ? $"{ecosystem}:{name}" : $"{ecosystem}:{group}:{name}";

    /// <summary>Resolve CALLS target symbol (qualified_name, display name) — ADR 0023 Rules 1–3.</summary>
    public static (string QualifiedName, string Name) ResolveCallTarget(
        string callToken,
        string collection,
        string relPath,
        IReadOnlyDictionary<string, List<GraphDefineEntry>> definesByName,
        IReadOnlyList<string> fileImports)
    {
        _ = relPath;

        if (definesByName.TryGetValue(callToken, out var exact) && exact.Count == 1)
        {
            var e = exact[0];
            return (e.QualifiedName, e.Name);
        }

        if (!callToken.Contains('.', StringComparison.Ordinal))
        {
            return (CalleeQualifiedName(collection, callToken), callToken);
        }

        var dot = callToken.LastIndexOf('.');
        var receiver = callToken[..dot];
        var method = callToken[(dot + 1)..];

        if (definesByName.TryGetValue(method, out var byMethod) && byMethod.Count == 1)
        {
            var e = byMethod[0];
            return (e.QualifiedName, e.Name);
        }

        var matchingImports = fileImports
            .Where(imp => string.Equals(imp, receiver, StringComparison.OrdinalIgnoreCase))
            .ToArray();
        if (matchingImports.Length > 0 && definesByName.TryGetValue(method, out var methodEntries))
        {
            var filtered = methodEntries
                .Where(e => matchingImports.Any(imp =>
                    e.RelPath.Contains(imp, StringComparison.OrdinalIgnoreCase)
                    || e.QualifiedName.Contains(imp, StringComparison.OrdinalIgnoreCase)))
                .ToArray();
            if (filtered.Length == 1)
            {
                return (filtered[0].QualifiedName, filtered[0].Name);
            }
        }

        return (CalleeQualifiedName(collection, callToken), callToken);
    }

    /// <summary>Best-effort HTTP method inference from endpoint definition content.</summary>
    public static string InferHttpMethod(string content)
    {
        var match = HttpMethodPattern().Match(content);
        if (!match.Success)
        {
            return string.Empty;
        }

        var token = match.Value.ToLowerInvariant();
        foreach (var method in new[] { "get", "post", "put", "delete", "patch" })
        {
            if (token.Contains(method, StringComparison.Ordinal))
            {
                return method.ToUpperInvariant();
            }
        }

        return string.Empty;
    }

    /// <summary>Build a graph batch from indexed chunks (grouped by file).</summary>
    public GraphBatch BuildGraphBatch(
        string collection,
        IReadOnlyList<Chunk> chunks,
        UrlExtractors urlExtractors,
        string workspacePath,
        IReadOnlyList<string> collectionNames)
    {
        var batch = new GraphBatch(collection);
        if (chunks.Count == 0)
        {
            return batch;
        }

        var byFile = chunks.GroupBy(c => c.RelPath, StringComparer.Ordinal)
            .ToDictionary(g => g.Key, g => (IReadOnlyList<Chunk>)g.ToArray(), StringComparer.Ordinal);

        var definesByName = new Dictionary<string, List<GraphDefineEntry>>(StringComparer.Ordinal);
        foreach (var chunk in chunks)
        {
            if (string.IsNullOrEmpty(chunk.SymbolName))
            {
                continue;
            }

            var qn = SymbolQualifiedName(collection, chunk.RelPath, chunk.SymbolName);
            if (!definesByName.TryGetValue(chunk.SymbolName, out var list))
            {
                list = [];
                definesByName[chunk.SymbolName] = list;
            }

            list.Add(new GraphDefineEntry(qn, chunk.RelPath, chunk.SymbolName, chunk.SymbolType));
        }

        var importsByFile = new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal);
        var seenFiles = new HashSet<string>(StringComparer.Ordinal);
        var seenBuildKeys = new HashSet<string>(StringComparer.Ordinal);

        foreach (var (relPath, fileChunks) in byFile)
        {
            var first = fileChunks[0];
            if (seenFiles.Add(relPath))
            {
                batch.Files.Add(new GraphFileRow(relPath, first.Language, first.FileSha256));
            }

            foreach (var chunk in fileChunks)
            {
                batch.Chunks.Add(new GraphChunkRow(chunk.Id.Value, relPath, chunk.StartLine, chunk.EndLine));

                if (!string.IsNullOrEmpty(chunk.SymbolName))
                {
                    var qn = SymbolQualifiedName(collection, relPath, chunk.SymbolName);
                    batch.Defines.Add(new GraphDefineRow(chunk.Id.Value, qn, chunk.SymbolName, chunk.SymbolType));
                }

                foreach (var callee in chunk.Callees)
                {
                    if (!importsByFile.ContainsKey(relPath))
                    {
                        var fileContent = ReadFileContent(workspacePath, relPath)
                            ?? string.Join('\n', fileChunks.Select(c => c.Content));
                        importsByFile[relPath] = ImportNameExtractor.ExtractFileImportNames(fileContent, first.Language);
                    }

                    var (qn, symName) = ResolveCallTarget(
                        callee, collection, relPath, definesByName, importsByFile[relPath]);
                    batch.Calls.Add(new GraphCallRow(chunk.Id.Value, qn, symName, callee));
                }

                foreach (var route in UrlExtractors.RoutePaths(chunk.Content, relPath))
                {
                    var method = InferHttpMethod(chunk.Content);
                    batch.Endpoints.Add(new GraphEndpointRow(route, method));
                    batch.DeclaresEndpoint.Add(new GraphDeclaresEndpointRow(chunk.Id.Value, route));
                }

                foreach (var path in urlExtractors.CodeUrls(chunk.Content))
                {
                    batch.Endpoints.Add(new GraphEndpointRow(path, string.Empty));
                    batch.HttpCalls.Add(new GraphHttpCallRow(chunk.Id.Value, path));
                }

                if (ConfigFilePattern().IsMatch(relPath))
                {
                    var (configPaths, _) = urlExtractors.ConfigUrls(chunk.Content);
                    foreach (var path in configPaths)
                    {
                        batch.Endpoints.Add(new GraphEndpointRow(path, string.Empty));
                        batch.Configures.Add(new GraphConfiguresRow(chunk.Id.Value, path));
                    }
                }
            }

            var fullContent = ReadFileContent(workspacePath, relPath)
                ?? string.Join('\n', fileChunks.Select(c => c.Content));
            foreach (var importName in ImportNameExtractor.ExtractFileImportNames(fullContent, first.Language))
            {
                batch.Imports.Add(new GraphImportRow(
                    relPath, ImportQualifiedName(collection, importName), importName));
            }

            if (BuildManifestDetector.IsBuildManifest(relPath))
            {
                var manifestContent = ReadFileContent(workspacePath, relPath) ?? fullContent;
                var deps = BuildDepExtractor.Extract(manifestContent, relPath);
                var matches = BuildDepCollectionMatcher.Match(deps, collectionNames, collection);
                var matchByArtifact = matches
                    .GroupBy(m => m.Artifact, StringComparer.Ordinal)
                    .ToDictionary(g => g.Key, g => g.First().MatchedCollection, StringComparer.Ordinal);

                foreach (var dep in deps)
                {
                    var key = ArtifactKey(dep.Group, dep.Artifact, dep.Ecosystem);
                    if (!seenBuildKeys.Add(key))
                    {
                        continue;
                    }

                    batch.BuildDeps.Add(new GraphBuildDepRow(
                        key, dep.Artifact, dep.Group, dep.Ecosystem, dep.Version, dep.Scope));
                    if (matchByArtifact.TryGetValue(dep.Artifact, out var target))
                    {
                        batch.ResolvesTo.Add(new GraphResolvesToRow(key, target));
                    }
                }
            }
        }

        return batch;
    }

    /// <summary>Map each chunk_id to neighbor Neo4j node keys (ADR 0002 Phase 2).</summary>
    public static IReadOnlyDictionary<string, IReadOnlyList<string>> GraphNodeIdsFromBatch(GraphBatch batch)
    {
        var collection = batch.Collection;
        var result = new Dictionary<string, List<string>>(StringComparer.Ordinal);
        var seen = new Dictionary<string, HashSet<string>>(StringComparer.Ordinal);

        void Add(string chunkId, string key)
        {
            if (!seen.TryGetValue(chunkId, out var bucket))
            {
                bucket = new HashSet<string>(StringComparer.Ordinal);
                seen[chunkId] = bucket;
            }

            if (!bucket.Add(key))
            {
                return;
            }

            if (!result.TryGetValue(chunkId, out var list))
            {
                list = [];
                result[chunkId] = list;
            }

            list.Add(key);
        }

        foreach (var row in batch.Defines)
        {
            Add(row.ChunkId, row.QualifiedName);
        }

        foreach (var row in batch.Calls)
        {
            Add(row.ChunkId, row.QualifiedName);
        }

        foreach (var row in batch.DeclaresEndpoint)
        {
            Add(row.ChunkId, $"{collection}:{row.Path}");
        }

        foreach (var row in batch.HttpCalls)
        {
            Add(row.ChunkId, $"{collection}:{row.Path}");
        }

        foreach (var row in batch.Configures)
        {
            Add(row.ChunkId, $"{collection}:{row.Path}");
        }

        var chunkIdsByFile = new Dictionary<string, List<string>>(StringComparer.Ordinal);
        foreach (var chunk in batch.Chunks)
        {
            if (!chunkIdsByFile.TryGetValue(chunk.RelPath, out var list))
            {
                list = [];
                chunkIdsByFile[chunk.RelPath] = list;
            }

            list.Add(chunk.ChunkId);
        }

        foreach (var row in batch.Imports)
        {
            if (!chunkIdsByFile.TryGetValue(row.RelPath, out var chunkIds))
            {
                continue;
            }

            foreach (var chunkId in chunkIds)
            {
                Add(chunkId, row.QualifiedName);
            }
        }

        return result.ToDictionary(
            kv => kv.Key,
            kv => (IReadOnlyList<string>)kv.Value,
            StringComparer.Ordinal);
    }

    private string? ReadFileContent(string workspacePath, string relPath)
    {
        var path = Path.Combine(workspacePath, relPath.Replace('\\', '/'));
        try
        {
            return File.ReadAllText(path);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            _logger.LogWarning(ex, "graph_writer_read_failed path={Path}", relPath);
            return null;
        }
    }

    [GeneratedRegex(
        @"@(?:Get|Post|Put|Delete|Patch|Request)Mapping|" +
        @"\[(?:Http(?:Get|Post|Put|Delete|Patch))\]|" +
        @"\.Map(Get|Post|Put|Delete|Patch)\(|" +
        @"(?:app|router)\.(get|post|put|delete|patch)\(",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)]
    private static partial Regex HttpMethodPattern();

    [GeneratedRegex(@"\.(ya?ml|json|properties|env|config)$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)]
    private static partial Regex ConfigFilePattern();
}
